dart/common/Extensible.h

```diff
@@ -0,0 +1,217 @@
+/// creating data structures that are meant to be extended. Ordinary copying
+/// does not work with extended data structures, because information from the
+/// derived classes will typically be lost during a copy. The Extensible class
+/// provides an interface for creating a new copy of an extended data structure,
+/// as well as copying information between two extended data structures of the
+/// same type.
+template <class T>
+class Extensible
+{
+public:
+
+  /// Default constructor
+  Extensible() = default;
+
+  /// Virtual destructor
+  virtual ~Extensible() = default;
+
+  /// Do not copy this class directly, use clone() or copy() instead
+  Extensible(const Extensible& doNotCopy) = delete;
+
+  /// Do not copy this class directly, use clone() or copy() instead
+  Extensible& operator=(const Extensible& doNotCopy) = delete;
+
+  /// Implement this function to allow your Extensible type to be copied safely.
+  virtual std::unique_ptr<T> clone() const = 0;
+
+  /// Copy the contents of anotherExtensible into this one.
+  virtual void copy(const T& anotherExtensible) = 0;
+};
+
+/// The ExtensibleMixer class is used to easily create an Extensible (such as
+/// Node::State) which simply takes an existing class (Mixin) and creates an
+/// Extensible that wraps it. This creates all the appropriate copy, move, and
+/// clone members, allowing you to follow the Rule Of Zero. You can also
+/// construct an instance in the exact same way that you would construct a Mixin
+/// instance.
+template <class T, class Mixin>
+class ExtensibleMixer : public T, public Mixin
+{
+public:
+
+  /// Default constructor. Uses the default constructor of Mixin
+  ExtensibleMixer();
+
+  /// Templated constructor. Uses whichever Mixin constructor is able to match
+  /// the arguments.
+  template <typename ... Args>
+  ExtensibleMixer(Args&&... args);
+
+  /// Constructs using a Mixin instance
+  ExtensibleMixer(const Mixin& mixin);
+
+  /// Constructs using a Mixin rvalue
+  ExtensibleMixer(Mixin&& mixin);
+
+  /// Copy constructor
+  ExtensibleMixer(const ExtensibleMixer<T, Mixin>& other);
+
+  /// Move constructor
+  ExtensibleMixer(ExtensibleMixer<T, Mixin>&& other);
+
+  /// Copy assignment operator that uses a Mixin instance
+  ExtensibleMixer& operator=(const Mixin& mixin);
+
+  /// Move assignment operator that uses a Mixin rvalue
+  ExtensibleMixer& operator=(Mixin&& mixin);
+
+  /// Copy assignment operator
+  ExtensibleMixer& operator=(const ExtensibleMixer& other);
+
+  /// Move assignment operator
+  ExtensibleMixer& operator=(ExtensibleMixer&& other);
+
+  // Documentation inherited
+  std::unique_ptr<T> clone() const override final;
+
+  // Documentation inherited
+  void copy(const T& other) override final;
+
+};
+
+/// MapHolder is a templated wrapper class that is used to allow maps of
+/// Addon::State and Addon::Properties to be handled in a semantically
+/// palatable way.
+template <typename MapType>
+class ExtensibleMapHolder final
```

Comment:
Also, I do some clever things to avoid allocating memory whenever possible. Suppose you already have two ``ExtensibleVector``s both filled up with ``Extensible`` objects and you want to copy the data from the objects in one of those vectors to the objects in the other vector without wiping clean and reallocating all the objects in the receiving vector. ``ExtensibleVector`` will do that automatically when you use the assignment operator. It will only allocate an object if a corresponding object isn't allocated already.